### PR TITLE
filter background layer by terrain

### LINKED ISSUE #1357: Extend background imagery source filtering to show only terrain layers
Currently the background imagery selection dialog (accessible from the (+) in the layer dialog) allows selection of either all background layers, or aerial imagery. With more and more terrain layers becoming available (for example hill shading) it would be nice to allow filtering just for these.

### PR DESCRIPTION
**Pull request by zedlabs**
adds radio button to filter background imagery layers by terrain
fixes #1357

### CODE REVIEW

**Review by simonpoole — CHANGES_REQUESTED** [review]
This really needs a short test of switching between the categories. See https://github.com/MarcusWolschon/osmeditor4android/blob/master/src/androidTest/java/de/blau/android/layer/LayerDialogTest.java for some examples, in particular backgroundLayerTest for getting the dialog.

I would simply switch between the three possible states and check the displayed content.

**Comment by zedlabs** [comment]
@simonpoole please take a look at this test, I'm not really sure if this was the expected result

File: src/androidTest/java/de/blau/android/TestUtils.java [hunk 1]
```diff
@@ -1243,4 +1243,16 @@ public static void clickSimpleButton(@NonNull UiDevice device) {
     public static UiObject getLock(@NonNull UiDevice device) {
         return device.findObject(new UiSelector().resourceId(device.getCurrentPackageName() + ":id/floatingLock"));
     }
+
+    /**
+     * Check for a layer name in a layer list
+     *
+     * @return true if element exists
+     */
+    public static boolean findInParentList(String[] parentList, String child) {
```
Comment (simonpoole) [inline src/androidTest/java/de/blau/android/TestUtils.java:1252]: Annotations please

File: src/androidTest/java/de/blau/android/layer/LayerDialogTest.java [hunk 2]
```diff
@@ -262,4 +264,35 @@ public void backgroundLayer() {
...
     }
+
+    /**
+     * Checks if filters are correct
+     */
+    @Test
+    public void layerFilter() {
+        assertTrue(TestUtils.clickResource(device, true, device.getCurrentPackageName() + ":id/layers", true));
+        assertTrue(TestUtils.clickButton(device, device.getCurrentPackageName() + ":id/add", true));
+        assertTrue(TestUtils.clickText(device, false, main.getString(R.string.menu_layers_add_backgroundlayer), true, false));
```
Comment (simonpoole) [inline src/androidTest/java/de/blau/android/layer/LayerDialogTest.java:275]: After this you should add an, not asserted, click on the OK button to click away the Tip dialog if it is displayed (which will happen if you are just running this test), see the backgroundLayer test.

File: src/androidTest/java/de/blau/android/layer/LayerDialogTest.java [hunk 3]
```diff
@@ -262,4 +264,35 @@ public void backgroundLayer() {
...
+
+        assertTrue(TestUtils.clickText(device, true, "All", true, false));
+
+        assertTrue(TestUtils.clickText(device, true, "Aerial imagery", true, false));
+        UiObject2 bingAerial = TestUtils.findObjectWithText(device, false, "Bing aerial imagery", 1000);
+        List<UiObject2> childrenAerial = bingAerial.getParent().getChildren();
+        assertTrue(TestUtils.findInParentList(aerialNames, childrenAerial.get(0).getText()));
+
+        assertTrue(TestUtils.clickText(device, true, "Terrain", true, false));
+        UiObject2 stamenTerrain = TestUtils.findObjectWithText(device, false, "Stamen Terrain", 1000);
```
Comment (simonpoole) [inline src/androidTest/java/de/blau/android/layer/LayerDialogTest.java:292]: This is a bit brittle as there is no guarantee that the Stamen Terrain layer is available (in general you can only rely on the couple of entries that are Vespucci specific to be around). Best simply add a layer of the Terrain category with TileLayerSource.addOrUpdateCustomLayer at the beginning of the test with an unique name and check if that then turns up.

**Review by simonpoole — CHANGES_REQUESTED** [review]
Running the test on an Emulator I noticed an UI glitch that needs to be looked at, see

grafik

grafik

See the inline comments the other small issues.

**Comment by zedlabs** [comment]
@simonpoole Please take a look

**Review by simonpoole — COMMENTED** [review]
Could you rebase on master? This is probably good to merge, and it makes things simpler to review too. Thanks.

**Comment by zedlabs** [comment]
@simonpoole done, also please have a look at #1363

**Comment by simonpoole** [comment]
Thanks. https://github.com/MarcusWolschon/osmeditor4android/pull/1363 in due course.

**Pull request merged**